wen-restart/src/wen_restart.rs

```diff
@@ -1,58 +1,279 @@
+    let mut last_voted_fork_slots_aggregate = LastVotedForkSlotsAggregate::new(
+        root_slot,
+        REPAIR_THRESHOLD,
+        root_bank.epoch_stakes(root_bank.epoch()).unwrap(),
+        &progress
+            .my_last_voted_fork_slots
+            .as_ref()
+            .unwrap()
+            .last_voted_fork_slots,
+        &cluster_info.id(),
+    );
+    let mut cursor = solana_gossip::crds::Cursor::default();
+    let mut is_full_slots = HashSet::new();
+    if let Some(aggregate_record) = &progress.last_voted_fork_slots_aggregate {
+        for (key_string, message) in &aggregate_record.received {
+            let _ = last_voted_fork_slots_aggregate.aggregate_from_record(key_string, message)?;
+        }
+    } else {
+        progress.last_voted_fork_slots_aggregate = Some(LastVotedForkSlotsAggregateRecord {
+            received: HashMap::new(),
+        });
+    }
+    loop {
+        if exit.load(Ordering::Relaxed) {
+            return Err("Exiting".into());
+        }
+        let start = timestamp();
+        for new_last_voted_fork_slots in cluster_info.get_restart_last_voted_fork_slots(&mut cursor)
+        {
+            let from = new_last_voted_fork_slots.from.to_string();
+            if let Some(record) =
+                last_voted_fork_slots_aggregate.aggregate(new_last_voted_fork_slots)
+            {
+                progress
+                    .last_voted_fork_slots_aggregate
+                    .as_mut()
+                    .unwrap()
+                    .received
+                    .insert(from, record);
+            }
+        }
+        let result = last_voted_fork_slots_aggregate.get_aggregate_result();
+        let mut filtered_slots: Vec<Slot>;
+        {
+            let my_bank_forks = bank_forks.read().unwrap();
+            filtered_slots = result
+                .slots_to_repair
+                .into_iter()
+                .filter(|slot| {
+                    if slot <= &root_slot || is_full_slots.contains(slot) {
+                        return false;
+                    }
+                    let is_full = my_bank_forks
+                        .get(*slot)
+                        .map_or(false, |bank| bank.is_frozen());
+                    if is_full {
+                        is_full_slots.insert(*slot);
+                    }
+                    !is_full
+                })
+                .collect();
+        }
+        filtered_slots.sort();
+        info!(
+            "Active peers: {} Slots to repair: {:?}",
+            result.active_percent, &filtered_slots
+        );
+        if filtered_slots.is_empty()
+            && result.active_percent > wait_for_supermajority_threshold_percent as f64
+        {
+            *wen_restart_repair_slots.write().unwrap() = vec![];
+            break;
+        }
+        {
+            *wen_restart_repair_slots.write().unwrap() = filtered_slots;
+        }
+        write_wen_restart_records(wen_restart_path, progress)?;
```

Comment:
Isn't this write handled by the caller in `wait_for_wen_restart()`? https://github.com/solana-labs/solana/blob/1ceda569f53e83d4d31836380d722a181b9ffd81/wen-restart/src/wen_restart.rs#L220

For less complicated code and cleaner testing, I think:
1. We leave all the record keeping into `pub fn wait_for_wen_restart()`, so `wen_restart_path` never needs to be passed into these state transition functions. This also means for testing we don't have to continually test for all errors that records are written to `wen_restart_path` (we only need one test for that), we just have to test that the inner state transition function returns the expected state. 
2. Each of the state transition functions called here https://github.com/solana-labs/solana/blob/1ceda569f53e83d4d31836380d722a181b9ffd81/wen-restart/src/wen_restart.rs#L197-L219 should return a custom error type. For instance `aggregate_restart_last_voted_fork_slots()` should have an `enum AggregateForksError()`. This achieves a few things:
* For debugging, we know where each of the errors originated
* For testing, we can test that we can recover from all possible expected errors.

Reply:
I originally put all book keeping in wait_for_wen_restart, but then realized that AggregateLastVotedForkSlots might span several hours when we wait for validators to join the restart, then I decided we want to write received LastVotedForkSlots inside this loop so we can speed up internal state population in case the validator restarts.

What's your take on this? We can of course rely on others sending us the Gossip messages as well.

Reply:
Another reason we do want to keep record of the Gossip messages we received is that we can maybe debug later why the restart didn't work.

Reply:
Instead of adding one error type per internal function, I added one WenRestartError for all functions in the crate, so we can do ?; everywhere. How does this look?

Reply:
> I originally put all book keeping in wait_for_wen_restart, but then realized that AggregateLastVotedForkSlots might span several hours when we wait for validators to join the restart, then I decided we want to write received LastVotedForkSlots inside this loop so we can speed up internal state population in case the validator restarts.

What's your take on this? We can of course rely on others sending us the Gossip messages as well.

You're totally right, let's keep the record writing inside the function then, and test those records are properly written

Reply:
> Instead of adding one error type per internal function, I added one WenRestartError for all functions in the crate, so we can do ?; everywhere. How does this look?

Think it's better to still organize per state transition for testing, see comment below